src/charm.py

```diff
@@ -276,13 +280,12 @@ def _set_workload_version(self) -> None:
         except synapse.APIError as exc:
             logger.debug("Cannot set workload version at this time: %s", exc)
 
-    @inject_charm_state
-    def _on_config_changed(self, _: ops.HookEvent, charm_state: CharmState) -> None:
-        """Handle changed configuration.
+    @validate_charm_state
+    def _on_config_changed(self, _: ops.HookEvent) -> None:
+        """Handle changed configuration."""
+        charm_state = self.build_charm_state()
+        MASConfiguration.validate(self)
```

Comment:
I like the idea of just validating instead of creating the charm state. Also splitting the charm state can be a good thing. However, now, we have this everywhere:
```
        charm_state = self.build_charm_state()
        MASConfiguration.validate(self)
....
        reconcile(charm_state)
```

How far are we from building the charm state in the reconcile and having most of the handlers empty? Is it realistic?

Reply:
I'm marking this as resolved, we can continue the discussion internally if needed!